src/ParallelAlgorithms/Events/ObserveDataBox.hpp

```diff
@@ -0,0 +1,159 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <map>
+#include <pup.h>
+#include <tuple>
+#include <vector>
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "Domain/Structure/ElementId.hpp"
+#include "IO/Observer/ObserverComponent.hpp"
+#include "IO/Observer/ReductionActions.hpp"
+#include "Options/String.hpp"
+#include "Parallel/DistributedObject.hpp"
+#include "Parallel/GlobalCache.hpp"
+#include "Parallel/Invoke.hpp"
+#include "Parallel/Reduction.hpp"
+#include "ParallelAlgorithms/EventsAndTriggers/Event.hpp"
+#include "Utilities/PrettyType.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace Events {
+
+namespace detail {
+
+struct map_add {
+  std::map<std::string, size_t> operator()(
+      std::map<std::string, size_t> map_1,
+      const std::map<std::string, size_t>& map_2) {
+    for (const auto& [key, value] : map_2) {
+      map_1.at(key) += value;
+    }
+    return map_1;
+  }
+};
+
+using ReductionType = Parallel::ReductionData<
+    // Time
+    Parallel::ReductionDatum<double, funcl::AssertEqual<>>,
+    // Map of total mem usage in MB per item in DataBoxes
+    Parallel::ReductionDatum<std::map<std::string, size_t>, map_add>>;
+
+template <typename ContributingComponent>
+struct ReduceDataBoxSize {
+  template <typename ParallelComponent, typename DbTags, typename Metavariables,
+            typename ArrayIndex>
+  static void apply(db::DataBox<DbTags>& /*box*/,
+                    Parallel::GlobalCache<Metavariables>& cache,
+                    const ArrayIndex& /*array_index*/, const double time,
+                    const std::map<std::string, size_t>& item_sizes) {
+    auto& observer_writer_proxy = Parallel::get_parallel_component<
+        observers::ObserverWriter<Metavariables>>(cache);
+    const std::string subfile_name =
+        "/ObserveDataBox/" + pretty_type::name<ContributingComponent>();
```

Comment:
Should we make this file name be specified in `ObserveDataBox` like we do for other observations? Maybe otherwise name this `ObserveDataBoxSizeInMb`? Long, but extremely unlikely to conflict with anything